README.md

```diff
@@ -421,6 +421,91 @@ BENCHMARK_REGISTER_F(MyFixture, BarTest)->Threads(2);
 /* BarTest is now registered */
 ```
 
+
+## User-defined counters
+
+You can add your own counters with user-defined names. The example below
+will add columns "Foo", "Bar" and "Baz" in its output:
+
+```c++
+static void UserCountersExample1(benchmark::State& state) {
+  double numFoos = 0, numBars = 0, numBazs = 0;
+  while (state.KeepRunning()) {
+    // ... count Foo,Bar,Baz events
+  }
+  state.counters["Foo"] = numFoos;
+  state.counters["Bar"] = numBars;
+  state.counters["Baz"] = numBazs;
+}
+```
+
+The `state.counters` object is a map-like container of `Counter` objects.
+Its `operator[](const char* name)` returns the counter with the given name or,
+if none is present, inserts and then returns a new `Counter` with that name.
+The `Counter` class has assignment (`=`) and compound assignment overloads
+(`+=`, `-=`, `*=`, `/=`) receiving a `double` as argument.
+
+`state.counters` also offers an `Insert()` method. This method can be used both
+to insert a new counter or update an existing counter:
+
+```c++
+  state.counters.Insert("name"); // inserts a new counter, with value of 0
+  double someValue;
+  state.counters.Insert("name", someValue); // inserts a new counter or updates an existing counter
+```
+
+An additional overload to `state.counters.Insert()` allows you to mark
+counters as rates and/or as per-thread averages:
+
+```c++
+  // Set the counter as a rate. It will be presented divided
+  // by the duration of the benchmark.
+  state.counters.Insert("FooRate", numFoos, benchmark::Counter::kIsRate);
+
+  // Set the counter as a thread-average quantity. It will
+  // be presented divided by the number of threads. */
+  state.counters.Insert("FooAvg", numFoos, benchmark::Counter::kAvgThreads);
+
+  // the third argument is a bitwise mask, so you can combine.
+  // This will mark as a thread-average rate:
+  state.counters.Insert("FooAvgRate", numFoos, benchmark::Counter::kIsRate|benchmark::Counter::kAvgThreads);
+```
+
+A further use of `state.counters.Insert()` is to minimize lookup time.
```

Comment:
given we know this can be an issue, and it's hard for users to know which to use (or they use the simplest, which is expensive) can we do anything to ensure they do the right thing? ie, should Insert be the only API and should we restrict it to only being called outside KeepRunning?